code/modules/power/supermatter/supermatter.dm

```diff
@@ -1,438 +1,876 @@
+		else
+			powerloss_dynamic_scaling = clamp(powerloss_dynamic_scaling - 0.05, 0, 1)
+		powerloss_inhibitor = clamp(1 - (powerloss_dynamic_scaling * clamp(combined_gas / POWERLOSS_INHIBITION_MOLE_BOOST_THRESHOLD, 1 , 1.5)), 0 , 1)
+
+		//Releases stored power into the general pool
+		//We get this by consuming shit or being scalpeled
+		if(matter_power && power_changes)
+			//We base our removed power off one 10th of the matter_power.
+			var/removed_matter = max(matter_power/MATTER_POWER_CONVERSION, 40)
+			//Adds at least 40 power
+			power = max(power + removed_matter, 0)
+			//Removes at least 40 matter power
+			matter_power = max(matter_power - removed_matter, 0)
+
+		var/temp_factor = 50
+		if(gasmix_power_ratio > 0.8)
+			// with a perfect gas mix, make the power less based on heat
+			icon_state = "[base_icon_state]_glow"
+		else
+			// in normal mode, base the produced energy around the heat
+			temp_factor = 30
+			icon_state = base_icon_state
+
+		if(power_changes)
+			power = max((removed.temperature * temp_factor / T0C) * gasmix_power_ratio + power, 0)
+
+		if(prob(50))
+			for(var/obj/machinery/power/rad_collector/R in GLOB.rad_collectors)
+				if(R.z == z && get_dist(R, src) <= 15) //Better than using orange() every process
+					R.receive_pulse(power * (1 + power_transmission_bonus / 10))
+
+		//Power * 0.55 * a value between 1 and 0.8
+		var/device_energy = power * REACTION_POWER_MODIFIER
+
+		//To figure out how much temperature to add each tick, consider that at one atmosphere's worth
+		//of pure oxygen, with all four lasers firing at standard energy and no N2 present, at room temperature
+		//that the device energy is around 2140. At that stage, we don't want too much heat to be put out
+		//Since the core is effectively "cold"
+
+		//Also keep in mind we are only adding this temperature to (efficiency)% of the one tile the rock
+		//is on. An increase of 4*C @ 25% efficiency here results in an increase of 1*C / (#tilesincore) overall.
+		//Power * 0.55 * (some value between 1.5 and 23) / 5
+		removed.temperature += ((device_energy * dynamic_heat_modifier) / THERMAL_RELEASE_MODIFIER)
+		//We can only emit so much heat, that being 57500
+		removed.temperature = max(0, min(removed.temperature, 2500 * dynamic_heat_modifier))
+
+		//Calculate how much gas to release
+		//Varies based on power and gas content
+		removed.toxins += max((device_energy * dynamic_heat_modifier) / PLASMA_RELEASE_MODIFIER, 0)
+		//Varies based on power, gas content, and heat
+		removed.oxygen += max(((device_energy + removed.temperature * dynamic_heat_modifier) - T0C) / OXYGEN_RELEASE_MODIFIER, 0)
+
+		if(produces_gas)
+			env.merge(removed)
+			air_update_turf()
+
+	//Makes em go mad and accumulate rads.
+	for(var/mob/living/carbon/human/l in view(src, HALLUCINATION_RANGE(power))) // If they can see it without mesons on.  Bad on them.
+		if(!istype(l.glasses, /obj/item/clothing/glasses/meson))
```

Comment:
I think this needs a check to make sure the affected mob can actually hallucinate (otherwise the IPCs will hallucinate)